### Import modules and DataFrame

In [1]:
import pandas as pd
import numpy as np
import pickle
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
dfs = pickle.load(open('dfs_churn_marked.p'))

In [3]:
#dfs.info()

### Categorical variables

In [4]:
unique_abtm = dfs['abtm'].unique() # A/B test mark
print len(unique_abtm)
print unique_abtm

1
[u'Release']


##### => Drop abtm

In [5]:
unique_br = dfs['br'].unique() # The bitrate of the media (kbps)
print len(unique_br)
print unique_br

19203
[u'3614101' u'3328416' u'3587620' ..., u'871378' u'2270765' u'2252671']


In [6]:
unique_brn = dfs['brn'].unique() # Browser name
print len(unique_brn)
print unique_brn

3
[u'KOOLi_Player' u'KOOLi_Humax_Player' nan]


In [7]:
unique_brv = dfs['brv'].unique() # Browser version
print len(unique_brv)
print unique_brv

9
[u'1.0.10' u'2.0.6' u'1.0.4' u'' nan u'2.0.7' u'1.0.7' u'2.0.4' u'1.0.11']


In [8]:
unique_chn = dfs['chn'].unique() # Live channel name
print len(unique_chn)
print unique_chn

33
[u'' u'CH.3 Gayo TV' u'CH.1 Yonhap News TV ' u'CH.16 CTS' u'CH.12 Edge TV'
 u'CH.14 The Movie' u'CH.17 FTV' u'CH.13 K-Baduk' u'CH.18 Mountain TV'
 u'CH.21 ONT' u'CH.26 GTV' u'CH.20 CBS' u'CH.28 Arirang TV' u'CH.31 CHING'
 u'CH.8 KBN TV' u'CH.15 Daekyo Kids TV' u'CH.10 Healthmedi TV' u'CH.33 ETN'
 u'CH.11 Runway TV' u'CH.23 Polaris TV' u'CH.22 Telenovela'
 u'CH.19 Channel China' u'CH.25 Food TV' u'CH.9 NTV' u'CH.34 KIDS-TV'
 u'CH.32 OBS W' u'CH.29 Arte TV' u'CH.7 Event TV' u'CH.24 Silver-i TV'
 u'CH.35 Jungchul TV' u'CH.27 Jei English' u'CH.30 BBS'
 u'CH.5 Asia Business News TV']


In [9]:
unique_co = dfs['co'].unique() # Contry name
print len(unique_co)
print unique_co

10
[u'Canada' u'United States' u'Republic of Korea' u'Guam' u'United Kingdom'
 u'Mexico' u'Vietnam' u'China' u'Singapore' u'Japan']


In [10]:
unique_coiso = dfs['coiso'].unique() # Country ISO Code
print len(unique_coiso)
print unique_coiso

10
[u'CA' u'US' u'KR' u'GU' u'GB' u'MX' u'VN' u'CN' u'SG' u'JP']


##### => Drop coiso 

In [11]:
unique_ct = dfs['ct'].unique() # City name
print len(unique_ct)
print unique_ct

1058
[u'Coquitlam' u'Los Angeles' u'Laguna Woods' ..., u'Spanaway'
 u'Charlestown' u'New Hyde Park']


In [12]:
unique_devm = dfs['devm'].unique() # device model name
print len(unique_devm)
print unique_devm

4
[u'WooriKooli W' u'h3' nan u'H5A2CUDQOT']


In [13]:
unique_devv = dfs['devv'].unique() # device vendor
print len(unique_devv)
print unique_devv

4
[u'Hutek' u'humaxdigital' nan u'Humaxdigital']


In [14]:
unique_eps = dfs['eps'].unique() # Episode name
print len(unique_eps)
print unique_eps

8093
[u"E71 : I'm Sorry Gangnamgu" u'E106 : Always Spring Day'
 u'E1133 : Open Concert' ..., u'E26 : You Are Gift' u'E306 : Running Man'
 u'E29 : Beautiful You']


In [15]:
unique_errmsg = dfs['errmsg'].unique() # Error message string
print len(unique_errmsg)
print unique_errmsg[:10]

246
[ u'com.google.android.exoplayer.audio.AudioTrack$InitializationException: AudioTrack init failed: 0, Config(44100, 12, 44100)'
 u'Unexpected IllegalStateException: null'
 u'java.net.SocketTimeoutException' nan
 u'Unable to connect to http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/sbs/20170401/6439023/video_und_2/19440000000.m4s'
 u'Unable to connect to http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/tvn/20170403/6439822/audio_und_1/9120107483.m4s'
 u'com.google.android.exoplayer.upstream.HttpDataSource$HttpDataSourceException: Unable to connect to http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/tvn/20170403/6439822/video_und_2/9120000000.m4s'
 u'Unable to connect to http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/tvn/20170403/6439822/6439822.mpd'
 u'Unable to connect to http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/sbs/20170402/6439609/video_und_5/2640000000.m4s'
 u'Unable to connect to http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/sbs/20170402/6439649/video_und_4/0.m4s

In [16]:
unique_gen = dfs['gen'].unique() # Gender
print len(unique_gen)
print unique_gen

2
[u'Unset' nan]


##### => Drop gen

In [17]:
unique_img = dfs['img'].unique() # Movie Thumbnail Image URL
print len(unique_img)
print unique_img

3073
[u'https://d1ijxi6ljrlc3i.cloudfront.net/5i7bozhpqk0fva6sxqrk'
 u'https://d1ijxi6ljrlc3i.cloudfront.net/hik9csyxhv60jhn7rgw9'
 u'https://d1ijxi6ljrlc3i.cloudfront.net/dedlbb7aygwme6zauxca' ...,
 u'https://d1ijxi6ljrlc3i.cloudfront.net/t0g9zajuyrxwgvax2t16'
 u'https://d1ijxi6ljrlc3i.cloudfront.net/tl23ima4g2qxnwwn6oxw'
 u'https://d1ijxi6ljrlc3i.cloudfront.net/h75urvro9zaa2ayjfslc']


In [18]:
unique_isp = dfs['isp'].unique() # ISP name
print len(unique_isp)
print unique_isp[:10]

114
[u'Pro Omnis Telecommunication' u'Time Warner Cable'
 u'West Coast Internet Incorporated' u'Comcast Cable' u'Verizon Fios'
 u'Frontier Communications Corporation' u'CenturyLink'
 u'AT&T Internet Services' u'AT&T U-verse' u'PenTeleData']


In [19]:
unique_lang = dfs['lang'].unique() # Language
print len(unique_lang)
print unique_lang

2
[u'ko' u'en']


In [20]:
unique_live = dfs['live'].unique() # live stream ('t') or not ('f')
print len(unique_live)
print unique_live

2
[u'f' u't']


In [21]:
unique_lrg = dfs['lrg'].unique() # Language Region and More
print len(unique_lrg)
print unique_lrg

3
[u'US' u'KR' u'Unset']


In [22]:
unique_mplyevnt = dfs['mplyevnt'].unique() # Player event type
print len(unique_mplyevnt)
print unique_mplyevnt

9
[u'iBufStart' u'brchg' u'compl' u'bufStart' u'stop' u'play' u'firstPlay'
 u'pause' nan]


In [23]:
unique_mpv = dfs['mpv'].unique() # Media player version
print len(unique_mpv)
print unique_mpv

8
[u'KOOLi_Player 1.0.10' u'KOOLi_Player 2.0.6' u'KOOLi_Player 1.0.4'
 u'KOOLi_Player ' u'KOOLi_Player 2.0.7' u'KOOLi_Player 1.0.7'
 u'KOOLi_Player 2.0.4' u'KOOLi_Player 1.0.11']


In [24]:
unique_mvcp = dfs['mvcp'].unique() # Movie contents provider (abc, ...)
print len(unique_mvcp)
print unique_mvcp

1
[u'Unset']


##### => Drop mvcp

In [25]:
unique_mvctg = dfs['mvctg'].unique() # Movie category (drama, ...)
print len(unique_mvctg)
print unique_mvctg

8
[u'Drama' u'Variety' u'Kids' u'News' u'Life' u'Movies' u'Live' u'LiveTV']


In [26]:
unique_mvid = dfs['mvid'].unique() # Movie ID
print len(unique_mvid)
print unique_mvid

8124
[u"E71 : I'm Sorry Gangnamgu" u'E106 : Always Spring Day'
 u'E1133 : Open Concert' ..., u'E26 : You Are Gift' u'E306 : Running Man'
 u'E29 : Beautiful You']


In [27]:
unique_mvrt = dfs['mvrt'].unique() # Movie rate(pg, r, or ...)
print len(unique_mvrt)
print unique_mvrt

1
[u'Unset']


##### => Drop mvrt

In [28]:
unique_mvsctg = dfs['mvsctg'].unique() # Movie sub-category(romantic,..)
print len(unique_mvsctg)
print unique_mvsctg

1
[u'Unset']


##### => Drop mvsctg

In [29]:
unique_osn = dfs['osn'].unique() # OS name
print len(unique_osn)
print unique_osn

2
[u'Android' nan]


In [30]:
unique_osv = dfs['osv'].unique() # OS version
print len(unique_osv)
print unique_osv

3
[u'4.4.4' u'4.4.2' nan]


In [31]:
unique_pltn = dfs['pltn'].unique() # Platform name
print len(unique_pltn)
print unique_pltn

3
[u'WooriKooli W' u'h3' u'H5A2CUDQOT']


In [32]:
unique_ppv = dfs['ppv'].unique() # Player Platform Version
print len(unique_ppv)
print unique_ppv

2
[u'Exoplayer:r1.4.2' u'Android Player']


In [33]:
unique_res = dfs['res'].unique() # Display resolutions
print len(unique_res)
print unique_res

2
[u'1080p' nan]


In [34]:
unique_rg = dfs['rg'].unique() # Region Name (State of province)
print len(unique_rg)
print unique_rg[:10]

65
[u'British Columbia' u'California' u'Massachusetts' u'New York' u'Ontario'
 u'New Jersey' u'Colorado' u'Illinois' u'Georgia' u'Pennsylvania']


In [35]:
unique_rgiso = dfs['rgiso'].unique() # Region ISO Code
print len(unique_rgiso)
print unique_rgiso[:10]

62
[u'BC' u'CA' u'MA' u'NY' u'ON' u'NJ' u'CO' u'IL' u'GA' u'PA']


In [36]:
unique_sid = dfs['sid'].unique() # Web server session ID
print len(unique_sid)
print unique_sid

3
[u'Wooli_KOOLi' u'KOOLi_Humax' u'KOOLi_Humax_Live']


In [37]:
unique_spos = dfs['spos'].unique() # Symbolic Position
print len(unique_spos)
print unique_spos[:10]

159
[u'V0009' u'V0034' u'V0058' u'V0000' u'V0022' u'V0061' u'V0039' u'V0006'
 u'V0038' u'V0013']


In [38]:
unique_srn = dfs['srn'].unique() # Series name
print len(unique_srn)
print unique_srn[:10]

470
[u'\uc544\uc784\uc3d8\ub9ac \uac15\ub0a8\uad6c'
 u'\uc5b8\uc81c\ub098 \ubd04\ub0a0' u'\uc5f4\ub9b0\uc74c\uc545\ud68c'
 u'\ud798\uc388\uc5ec\uc790 \ub3c4\ubd09\uc21c'
 u'\uc751\ub2f5\ud558\ub77c 1988' u'Tunnel' u'\uae40\uacfc\uc7a5'
 u'\uc625\uc911\ud654' u'\ubfcc\ub9ac\uae4a\uc740 \ub098\ubb34'
 u'\uac11\ub3d9\uc774 ']


In [39]:
unique_ssid = dfs['ssid'].unique() # Session ID
print len(unique_ssid)
print unique_ssid[:10]

58827
[u'be86c466-c943-4f35-86f0-e7dbe678def0'
 u'894819a2-ca4a-47d9-9e5c-6bd2ce54277e'
 u'e7f78970-b677-4655-a98d-00c5dcb4f6b0'
 u'2ff7c2f2-47b8-449c-8e80-8d7e6b4e0892'
 u'0c7169ce-6bff-4758-8f65-78728f361bc8'
 u'c592de2d-c731-40bd-bbe4-b3bee2702945'
 u'6f773d36-8a6f-4984-ba69-5bcb054e3b57'
 u'b92e852d-2e09-4655-a786-a9b469e43ed8'
 u'12ee6940-e994-43e6-9e29-7ff36c749f4e'
 u'acbbbc5d-a273-4bcb-93bd-64024271c597']


In [40]:
unique_stype = dfs['stype'].unique() # Service Type(TV, Radio or ...)
print len(unique_stype)
print unique_stype

1
[u'Unset']


##### => Drop stype

In [41]:
unique_svr = dfs['svr'].unique() # Server name or Server URL
print len(unique_svr)
print unique_svr[:10]

8351
[ u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/sbs/20170331/6437020/6437020.mpd'
 u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/mbc/20170330/6435628/6435628.mpd'
 u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/kbs1/20170101/6345142/6345142.mpd'
 u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/jtbc/20170331/6438161/6438161.mpd'
 u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/tvn/20160808/6218295/6218295.mpd'
 u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/tvn/20170326/6431923/6431923.mpd'
 u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/sbs/20170320/6425408/6425408.mpd'
 u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/kbs2/20170331/6436752/6436752.mpd'
 u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/mbc/20161106/6298460/6298460.mpd'
 u'http://kooli-cdn.wecandeo.com/kooli.wecandeo.com/sbs/20160922/6253443/6253443.mpd']


In [42]:
unique_szrid = dfs['szrid'].unique() # STZR Unique ID
print len(unique_szrid)
print unique_szrid[:10]

4359
[u'57d80ee3-4563-40c1-a64d-15a66c60edba'
 u'dc742456-10a3-48a5-9d4d-d8e8b220b069'
 u'95af6832-5e9d-4b07-91c3-7ef5a863e014'
 u'a2cf4b4f-7c7d-42df-87c3-96b2a5debf7a'
 u'f5d5c858-cc2f-4dec-831e-63fd6bb539a6'
 u'd6eaf52a-022d-4ad1-9cc5-68bdae59cfb1'
 u'd3d51af6-1eed-421e-852a-179b5f4ef83d'
 u'da713b93-6eec-4f54-8c0a-9b431343f1a9'
 u'363645eb-5557-444a-9eac-dd4a5205cd69'
 u'07cc27c4-4db3-41a1-85b8-1bc20bc0be84']


In [43]:
unique_tid = dfs['tid'].unique() # ticket ID
print len(unique_tid)
print unique_tid[:10]

64727
[u'4e7475b2-0b88-461f-9387-0a9b77251df8'
 u'7e588404-453f-4a45-a8aa-05b6b0ae5be4'
 u'f19a8414-64cd-418a-985e-61ae073d8915'
 u'7efccc8b-436d-40bf-8bd5-fe45d08039ca'
 u'14bd5375-4a03-460c-b0aa-d73fe16628b7'
 u'fb5b6348-3f14-4f8c-a5fe-f2633f20de40'
 u'8d04439a-03ea-416a-a3c6-4b2c008a6ce3'
 u'2c25e4ee-48bb-41c8-a167-f28585a0e106'
 u'468cb25d-92fa-43d5-8295-af5a4552c00e'
 u'd42e2d6e-5c75-43b2-ae4d-f4bd69ca3b2a']


In [44]:
unique_type = dfs['type'].unique() # log type
print len(unique_type)
print unique_type

8
[u'vh' u'lBuf' u'mplyevent' u'iBuf' u'upImg' u'errmsg' u'plyRdy' u'Buf']


In [45]:
unique_uid = dfs['uid'].unique() # User id
print len(unique_uid)
print unique_uid[:10]

4197
[u'WUCH155103440' u'14110963601665' u'WUCH155102826' u'WUCH161600388'
 u'WUCH155101681' u'WUCH155102162' u'WUCH155101131' u'WUCH155101270'
 u'WUCH155102245' u'WUCH155102068']


In [46]:
unique_utype = dfs['utype'].unique() # User type
print len(unique_utype)
print unique_utype

1
[u'user']


##### => Drop utype

In [47]:
unique_v = dfs['v'].unique() # Client API version
print len(unique_v)
print unique_v

1
[u'0.9.1']


##### => Drop v

In [48]:
unique_yob = dfs['yob'].unique() # Year of birth
print len(unique_yob)
print unique_yob

2
[u'Unset' nan]


In [49]:
unique_ts2 = dfs['ts2'].unique() # timestamp
print len(unique_ts2)
print unique_ts2

180853
['2017-03-31T23:56:07.535000000' '2017-03-31T23:57:12.602000000'
 '2017-03-31T23:56:57.448000000' ..., '2017-06-29T23:58:45.183000000'
 '2017-06-29T23:58:22.515000000' '2017-06-29T23:58:45.635000000']


### Now, drop them

In [50]:
def dropping_col(dfs, col_name):
    if col_name in dfs.columns:
        dfs = dfs.drop(col_name, axis = 1)
    return dfs

drop_col_lst = ['abtm','coiso','gen','mvcp','mvrt','mvsctg','stype','utype','v']
for col in drop_col_lst:
    dfs = dropping_col(dfs,col)

In [51]:
#dfs.info()

# 08/15/2017 Understanding logs..

    1. Groupby tid, and inspect
    2. Groupby szrid, and inspect
    3. Groupby uid, and inspect

### Dropping more columns...

In [52]:
# Only columns of interest
tmp_dfs = dfs[['ts2','drt','errmsg','mplyevnt','spos','ssid','szrid','tBuf','tIBuf','tLBuf','tPld','tVH','tid','type','uid','churn']]

In [53]:
tmp_dfs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 182000 entries, 0 to 181999
Data columns (total 16 columns):
ts2         182000 non-null datetime64[ns]
drt         182000 non-null int64
errmsg      2262 non-null object
mplyevnt    41523 non-null object
spos        167826 non-null object
ssid        182000 non-null object
szrid       182000 non-null object
tBuf        182000 non-null int64
tIBuf       182000 non-null int64
tLBuf       182000 non-null int64
tPld        182000 non-null int64
tVH         182000 non-null int64
tid         182000 non-null object
type        182000 non-null object
uid         182000 non-null object
churn       182000 non-null bool
dtypes: bool(1), datetime64[ns](1), int64(6), object(8)
memory usage: 22.4+ MB


### Groupby tid

In [54]:
tmp_dfs.groupby('tid').count().iloc[:10,:]
# Not useful..

ts2  drt  errmsg  mplyevnt  spos  ssid  \
tid                                                                            
0000dc43-0653-4dd3-be67-1710eddbbef2    2    2       0         1     2     2   
00012b82-1ab4-4109-a1f0-53ac7073db93    3    3       0         0     3     3   
00034ba6-ccb3-4c22-9af3-5a7df3d991cf    3    3       0         2     3     3   
00044168-4c0a-4b80-b3a7-9804427c3cc1    6    6       0         1     4     6   
00056954-6e24-4135-8dc9-98937506358a    1    1       0         0     1     1   
00061daf-d2d7-442c-9008-c81991e2e0f8    1    1       0         0     1     1   
00070c33-1bea-47f9-89d0-d49e05f9e30e    6    6       0         3     5     6   
0008596c-3f2d-4fb4-b80e-0bff2a915169    1    1       0         0     1     1   
00098dda-9e71-4e47-a144-4046032ebb59    1    1       0         0     1     1   
000b107f-952a-4082-9153-8938ea3576d2    3    3       0         0     3     3   

                                      szrid  tBuf  tIBuf  tLBuf  tPld  tVH  \
tid                                                                          
0000dc43-0653-4dd3-be67-1710eddbbef2      2     2      2      2     2    2   
00012b82-1ab4-4109-a1f0-53ac7073db93      3     3      3      3     3    3   
00034ba6-ccb3-4c22-9af3-5a7df3d991cf      3     3      3      3     3    3   
00044168-4c0a-4b80-b3a7-9804427c3cc1      6     6      6      6     6    6   
00056954-6e24-4135-8dc9-98937506358a      1     1      1      1     1    1   
00061daf-d2d7-442c-9008-c81991e2e0f8      1     1      1      1     1    1   
00070c33-1bea-47f9-89d0-d49e05f9e30e      6     6      6      6     6    6   
0008596c-3f2d-4fb4-b80e-0bff2a915169      1     1      1      1     1    1   
00098dda-9e71-4e47-a144-4046032ebb59      1     1      1      1     1    1   
000b107f-952a-4082-9153-8938ea3576d2      3     3      3      3     3    3   

                                      type  uid  churn  
tid                                                     
0000dc43-0653-4dd3-be67-1710eddbbef2     2    2      2  
00012b82-1ab4-4109-a1f0-53ac7073db93     3    3      3  
00034ba6-ccb3-4c22-9af3-5a7df3d991cf     3    3      3  
00044168-4c0a-4b80-b3a7-9804427c3cc1     6    6      6  
00056954-6e24-4135-8dc9-98937506358a     1    1      1  
00061daf-d2d7-442c-9008-c81991e2e0f8     1    1      1  
00070c33-1bea-47f9-89d0-d49e05f9e30e     6    6      6  
0008596c-3f2d-4fb4-b80e-0bff2a915169     1    1      1  
00098dda-9e71-4e47-a144-4046032ebb59     1    1      1  
000b107f-952a-4082-9153-8938ea3576d2     3    3      3

### Let's see what are those..

In [55]:
tid_grouped = tmp_dfs.groupby('tid')

In [56]:
type(tid_grouped.groups)

dict

In [57]:
tid_grouped.groups.keys()[:5]

[u'6fac8ee6-69c6-4a92-851b-9e0a055694bb',
 u'9cf0b8cf-3e38-4324-8b39-e4133a9d83c7',
 u'6c9305ea-fe8e-4252-8b89-e5724dc3a6f5',
 u'b83c45a3-5f4b-4d83-a4a5-d5e2eac6aec6',
 u'c48656f1-42d3-4f82-8be9-5def0b03a26c']

In [58]:
# How many 'tid's
print len(tid_grouped.groups.keys())

64727


In [59]:
tid_lst = tmp_dfs['tid'].unique()
print 'tid_lst: ',len(tid_lst)

tid_lst:  64727


In [60]:
# Sample
for tid in tid_lst[:2]:
    print '<tid>: ',tid
    print ''
    print '<tid dataframe>:',tid_grouped.get_group(tid)
    print type(tid_grouped.get_group(tid))
    print '=='*40
    print ''

<tid>:  4e7475b2-0b88-461f-9387-0a9b77251df8

<tid dataframe>:                       ts2      drt  \
0 2017-03-31 23:56:07.535  2234000   

                                              errmsg   mplyevnt   spos  \
0  com.google.android.exoplayer.audio.AudioTrack$...  iBufStart  V0009   

                                   ssid                                 szrid  \
0  be86c466-c943-4f35-86f0-e7dbe678def0  57d80ee3-4563-40c1-a64d-15a66c60edba   

   tBuf  tIBuf  tLBuf   tPld    tVH                                   tid  \
0   821      6   5003  49377  59880  4e7475b2-0b88-461f-9387-0a9b77251df8   

  type            uid  churn  
0   vh  WUCH155103440  False  
<class 'pandas.core.frame.DataFrame'>

<tid>:  7e588404-453f-4a45-a8aa-05b6b0ae5be4

<tid dataframe>:                          ts2      drt  \
1    2017-03-31 23:57:12.602  5974000   
38   2017-03-31 23:57:17.607  1512000   
510  2017-03-31 23:57:27.618  3198000   
1137 2017-03-31 23:57:47.640        0   
1284 2017-03-31 23:57:52

    1. It is sorted by timestamp which is nice
    2. 'ssid', 'szrid', 'uid' could be same... Really?
    3. How about 'type'?

In [61]:
# let's see
i=0;j=0;
for tid in tid_lst:
    if all(tid_grouped.get_group(tid).apply(lambda x: min(x) == max(x))[['ssid','szrid','uid','churn']]):
        i+=1;
    else: 
        print 'tid with different_ssid_szrid_uid_churn: ',tid
    if all(tid_grouped.get_group(tid).apply(lambda x: min(x) == max(x))[['type']]):
        j+=1;
print 'The number of tid with same ssid, szrid, uid and churn: ',i
print 'The number of tid with same type: ',j
print 'The number of unique tid: ',len(tid_lst)
print i == len(tid_lst) # Every tid has this
print j == len(tid_lst) # type are not same

tid with different_ssid_szrid_uid_churn:  Unset
The number of tid with same ssid, szrid, uid and churn:  64726
The number of tid with same type:  46537
The number of unique tid:  64727
False
False


    <same ssid, szrid, uid and churn>
    1. Off by 1 (# tid unique: 64727 vs 64726: # tid with same ssid, szrid, uid and churn)
    2. Find that group with tid == 'Unset'
    
    <same type>
    1. Off by a lot...
    2. Thus, type can vary within a tid

In [62]:
tid_grouped.get_group('Unset').iloc[:10,:] # group with tid == Unset

ts2      drt errmsg  mplyevnt   spos  \
153  2017-03-31 23:55:07.460  2679000    NaN  bufStart  V0037   
201  2017-03-31 23:55:07.463  3533000    NaN  bufStart  V0000   
594  2017-03-31 23:56:22.962  2897000    NaN       NaN  V0066   
682  2017-03-31 23:57:31.873        0    NaN       NaN  V0046   
1148 2017-03-31 23:55:41.422        0    NaN       NaN  V0003   
1378 2017-03-31 23:57:58.293        0    NaN       NaN  V0050   
1460 2017-03-31 23:57:58.296        0    NaN       NaN  V0063   
1604 2017-03-31 23:54:55.674        0    NaN       NaN  V0038   
1872 2017-03-31 23:58:15.032        0    NaN       NaN    NaN   
1962 2017-03-31 23:58:15.029        0    NaN       NaN    NaN   

                                      ssid  \
153   1a52f33a-af63-4787-bd1c-6ada10dd102f   
201   1a52f33a-af63-4787-bd1c-6ada10dd102f   
594   02b7e66a-d6d3-4f0a-a228-0aa3214496e0   
682   dc380759-21a3-4d49-8158-7414ebd4496d   
1148  084a3f86-7ea8-4266-8b64-3527a5fd0fcf   
1378  dc537617-63f9-49c5-a990-7cdf6904ef56   
1460  dc537617-63f9-49c5-a990-7cdf6904ef56   
1604  ad6a4964-43b7-41e7-9c95-950ac4b931ba   
1872  3ea2cd47-62fb-4dd2-a17e-097b39e4ce80   
1962  3ea2cd47-62fb-4dd2-a17e-097b39e4ce80   

                                     szrid  tBuf  tIBuf  tLBuf  tPld    tVH  \
153   387fb0a0-667b-4e32-acb9-ba37aa7841b0     0      0   5004     0  10000   
201   387fb0a0-667b-4e32-acb9-ba37aa7841b0     0      0   5004     0  59867   
594   a6ce7a11-3b65-4e6c-b048-5a1af639f197     0      0      0     0  32010   
682   6edf787c-6c02-4537-b8b1-9aee1d4d2c58     0      0      0     0      0   
1148  4fd26cd7-f05b-4b3c-a0a6-c65d890bd57e     0      0      0     0      0   
1378  44eb3f13-81be-4f60-b84f-6443fbd9dab8     0      0      0     0      0   
1460  44eb3f13-81be-4f60-b84f-6443fbd9dab8     0      0      0     0      0   
1604  914e6039-ed81-414a-add6-0092c62eb0b5     0      0      0     0      0   
1872  c351c96e-adf1-4d3d-82fe-6fe25493f3ff     0      0      0     0      0   
1962  c351c96e-adf1-4d3d-82fe-6fe25493f3ff     0      0      0     0      0   

        tid       type             uid  churn  
153   Unset     plyRdy   WUCH155101122  False  
201   Unset  mplyevent   WUCH155101122  False  
594   Unset  mplyevent   WUCH155100177   True  
682   Unset  mplyevent  14110962300019  False  
1148  Unset     plyRdy   WUCH155102898  False  
1378  Unset     plyRdy   WUCH155104392  False  
1460  Unset  mplyevent   WUCH155104392  False  
1604  Unset     plyRdy   WUCH155102391  False  
1872  Unset  mplyevent   WUCH155100947  False  
1962  Unset     plyRdy   WUCH155100947  False

    1. Many different ssid, szrid, uid and churn with the group with tid == 'Unset'
    2. It's better to remove this group with tid == 'Unset'

### Drop tid == 'Unset'

In [63]:
print 'whole: ',tmp_dfs.shape
print 'tid==Unset: ',tmp_dfs[tmp_dfs['tid']=='Unset'].shape
print 'after removing tid==Unset: ', tmp_dfs.shape[0] - tmp_dfs[tmp_dfs['tid']=='Unset'].shape[0]

whole:  (182000, 16)
tid==Unset:  (2253, 16)
after removing tid==Unset:  179747


In [64]:
tmp_dfs_rmv_tid_Uns = tmp_dfs[tmp_dfs['tid']!='Unset']

In [65]:
tmp_dfs_rmv_tid_Uns.shape 

(179747, 16)

In [66]:
tid_rmv_tid_Uns_grouped = tmp_dfs_rmv_tid_Uns.groupby('tid') 

# How many 'tid's
print len(tid_rmv_tid_Uns_grouped.groups.keys())

# How many 'tid's
tid_rmv_tid_Uns_lst = tmp_dfs_rmv_tid_Uns['tid'].unique()
print 'tid_rmv_tid_Uns_lst: ',len(tid_rmv_tid_Uns_lst)

64726
tid_rmv_tid_Uns_lst:  64726


In [67]:
# let's see
i=0;j=0;
for tid in tid_rmv_tid_Uns_lst:
    if all(tid_rmv_tid_Uns_grouped.get_group(tid).apply(lambda x: min(x) == max(x))[['ssid','szrid','uid','churn']]):
        i+=1;
    else: 
        print 'tid with different_ssid_szrid_uid_churn: ',tid
    if all(tid_rmv_tid_Uns_grouped.get_group(tid).apply(lambda x: min(x) == max(x))[['type']]):
        j+=1;
print 'The number of tid with same ssid, szrid, uid and churn: ',i
print 'The number of tid with same type: ',j
print 'The number of unique tid: ',len(tid_rmv_tid_Uns_lst)
print i == len(tid_rmv_tid_Uns_lst) # Every tid has this
print j == len(tid_rmv_tid_Uns_lst) # type are not same

The number of tid with same ssid, szrid, uid and churn:  64726
The number of tid with same type:  46537
The number of unique tid:  64726
True
False


In [68]:
# Sample
for tid in tid_rmv_tid_Uns_lst[0:2]:
    print '<tid>: ',tid
    print ''
    print '<tid dataframe>:',tid_rmv_tid_Uns_grouped.get_group(tid)
    print type(tid_rmv_tid_Uns_grouped.get_group(tid))
    print '=='*40
    print ''

<tid>:  4e7475b2-0b88-461f-9387-0a9b77251df8

<tid dataframe>:                       ts2      drt  \
0 2017-03-31 23:56:07.535  2234000   

                                              errmsg   mplyevnt   spos  \
0  com.google.android.exoplayer.audio.AudioTrack$...  iBufStart  V0009   

                                   ssid                                 szrid  \
0  be86c466-c943-4f35-86f0-e7dbe678def0  57d80ee3-4563-40c1-a64d-15a66c60edba   

   tBuf  tIBuf  tLBuf   tPld    tVH                                   tid  \
0   821      6   5003  49377  59880  4e7475b2-0b88-461f-9387-0a9b77251df8   

  type            uid  churn  
0   vh  WUCH155103440  False  
<class 'pandas.core.frame.DataFrame'>

<tid>:  7e588404-453f-4a45-a8aa-05b6b0ae5be4

<tid dataframe>:                          ts2      drt  \
1    2017-03-31 23:57:12.602  5974000   
38   2017-03-31 23:57:17.607  1512000   
510  2017-03-31 23:57:27.618  3198000   
1137 2017-03-31 23:57:47.640        0   
1284 2017-03-31 23:57:52

    1. O.K. Now, 'ssid','szrid','uid' are all same
    2. 'type' is not the same

### Groupby uid

In [69]:
uid_grouped = tmp_dfs.groupby('uid')
print uid_grouped.groups.keys()[:5]

[u'WUCH155102022', u'WUCH155102023', u'WUCH155102027', u'WUCH155102025', u'WUCH155102028']


In [70]:
# How many 'uid's
print len(uid_grouped.groups.keys())

4197


In [71]:
uid_lst = tmp_dfs['uid'].unique()
print 'uid_lst: ',len(uid_lst)

uid_lst:  4197


In [72]:
# Sample
for uid in uid_lst[:1]:
    print '<uid>: ',uid
    print ''
    print '<uid dataframe>:',uid_grouped.get_group(uid)
    print type(uid_grouped.get_group(uid))
    print '=='*40
    print ''

<uid>:  WUCH155103440

<uid dataframe>:                            ts2      drt  \
0      2017-03-31 23:56:07.535  2234000   
8659   2017-04-04 23:59:10.832  3169000   
10779  2017-04-05 23:58:55.037        0   
19644  2017-04-09 23:44:53.721        0   
38462  2017-04-19 23:49:18.169  5066000   
45519  2017-04-22 23:46:04.536        0   
89330  2017-05-14 23:48:42.355        0   
97491  2017-05-18 23:49:16.120        0   
106051 2017-05-23 23:50:50.805  3049000   
109840 2017-05-24 23:52:04.730        0   
122622 2017-05-31 23:48:22.227        0   
127633 2017-06-02 23:47:50.102        0   
127652 2017-06-02 23:47:50.310        0   
127664 2017-06-02 23:47:49.299        0   
127683 2017-06-02 23:47:50.307        0   
137958 2017-06-07 23:50:34.925        0   
166037 2017-06-22 23:52:50.786  4863000   
166681 2017-06-22 23:53:18.798        0   
166725 2017-06-22 23:53:18.796        0   
166806 2017-06-22 23:53:22.801        0   
166831 2017-06-22 23:53:22.805        0   
167426 2017-06

In [73]:
# let's see
i=0;j=0;
for uid in uid_lst:
    if all(uid_grouped.get_group(uid).apply(lambda x: min(x) == max(x))[['szrid']]):
        i+=1;
    else: 
        #print 'uid with different_szrid: ',uid
        #print 'diff szrids: ',uid_grouped.get_group(uid)['szrid'].unique()  # diff szrids
        #print ''
        j+=1;
print ''
print 'The number of uid with same szrid: ',i
print 'The number of uid with diff szrid: ',j
print 'The number of unique uid: ',len(uid_lst)
print (i+j) == len(uid_lst) 


The number of uid with same szrid:  4063
The number of uid with diff szrid:  134
The number of unique uid:  4197
True


In [74]:
# let's see
i=0;j=0;
for uid in uid_lst:
    if all(uid_grouped.get_group(uid).apply(lambda x: min(x) == max(x))[['churn']]):
        i+=1;
    else: 
        print 'uid with different_churn: ',uid
        print 'diff churn: ',uid_grouped.get_group(uid)['churn'].unique()  # diff szrids
        j+=1;
print ''
print 'The number of uid with same churn: ',i
print 'The number of uid with diff churn: ',j
print 'The number of unique uid: ',len(uid_lst)
print (i+j) == len(uid_lst) 


The number of uid with same churn:  4197
The number of uid with diff churn:  0
The number of unique uid:  4197
True


In [75]:
# let's see
i=0;j=0;
for uid in uid_lst:
    if all(uid_grouped.get_group(uid).apply(lambda x: min(x) == max(x))[['tid']]):
        i+=1;
    else: 
        #print 'uid with different_tid: ',uid
        #print 'diff tids: ',uid_grouped.get_group(uid)['tid'].unique()  # diff tids
        #print ''
        j+=1;
print ''
print 'The number of uid with same tid: ',i
print 'The number of uid with diff tid: ',j
print 'The number of unique uid: ',len(uid_lst)
print (i+j) == len(uid_lst) 


The number of uid with same tid:  293
The number of uid with diff tid:  3904
The number of unique uid:  4197
True


### Groupby szrid

In [76]:
szrid_grouped = tmp_dfs.groupby('szrid')

In [77]:
szrid_grouped.groups.keys()[:5]

[u'fa11475f-7d67-48f9-b6d8-cbf58c2601a4',
 u'f241da71-a260-492e-a151-becde3be99d6',
 u'78bfe5c0-9efd-4b23-ab77-97a8f0deded6',
 u'ed8e8bc6-4925-4e6d-a685-fde061e84487',
 u'f30eafa6-9e11-496c-b339-c8eba2953c7b']

In [78]:
# How many 'szrid's
print len(szrid_grouped.groups.keys())

4359


In [79]:
szrid_lst = tmp_dfs['szrid'].unique()
print 'szrid_lst: ',len(szrid_lst)

szrid_lst:  4359


In [80]:
# Sample
for szrid in szrid_lst[:1]:
    print '<szrid>: ',szrid
    print ''
    print '<szrid dataframe>:',szrid_grouped.get_group(szrid)
    print type(szrid_grouped.get_group(szrid))
    print '=='*40
    print ''

<szrid>:  57d80ee3-4563-40c1-a64d-15a66c60edba

<szrid dataframe>:                            ts2      drt  \
0      2017-03-31 23:56:07.535  2234000   
8659   2017-04-04 23:59:10.832  3169000   
10779  2017-04-05 23:58:55.037        0   
19644  2017-04-09 23:44:53.721        0   
38462  2017-04-19 23:49:18.169  5066000   
45519  2017-04-22 23:46:04.536        0   
89330  2017-05-14 23:48:42.355        0   
97491  2017-05-18 23:49:16.120        0   
106051 2017-05-23 23:50:50.805  3049000   
109840 2017-05-24 23:52:04.730        0   
122622 2017-05-31 23:48:22.227        0   
127633 2017-06-02 23:47:50.102        0   
127652 2017-06-02 23:47:50.310        0   
127664 2017-06-02 23:47:49.299        0   
127683 2017-06-02 23:47:50.307        0   
137958 2017-06-07 23:50:34.925        0   
166037 2017-06-22 23:52:50.786  4863000   
166681 2017-06-22 23:53:18.798        0   
166725 2017-06-22 23:53:18.796        0   
166806 2017-06-22 23:53:22.801        0   
166831 2017-06-22 23:53:22.805

In [81]:
# let's see
i=0;j=0;
for szrid in szrid_lst:
    if all(szrid_grouped.get_group(szrid).apply(lambda x: min(x) == max(x))[['uid']]):
        i+=1;
    else: 
        print 'szrid with different_uid: ',szrid
        print 'diff uids: ',szrid_grouped.get_group(szrid)['uid'].unique()  # diff uids
        j+=1;
print ''
print 'The number of szrid with same uid: ',i
print 'The number of szrid with diff uid: ',j
print 'The number of unique szrid: ',len(szrid_lst)
print (i+j) == len(szrid_lst) 


The number of szrid with same uid:  4359
The number of szrid with diff uid:  0
The number of unique szrid:  4359
True


In [82]:
# let's see
i=0;j=0;
for szrid in szrid_lst:
    if all(szrid_grouped.get_group(szrid).apply(lambda x: min(x) == max(x))[['tid']]):
        i+=1;
    else: 
        #print 'szrid with different_tid: ',uid
        #print 'diff tids: ',uid_grouped.get_group(uid)['tid'].unique()  # diff tids
        #print ''
        j+=1;
print ''
print 'The number of szrid with same tid: ',i
print 'The number of szrid with diff tid: ',j
print 'The number of unique szrid: ',len(szrid_lst)
print (i+j) == len(szrid_lst) 


The number of szrid with same tid:  347
The number of szrid with diff tid:  4012
The number of unique szrid:  4359
True


# Conclusion

### Groupby tid
    The number of tid with same ssid, szrid, uid and churn:  64726
    The number of unique tid:  64727
    tid with different_ssid_szrid_uid_churn:  Unset

    => Remove tid 'Unset'

    The number of tid with same ssid, szrid, uid and churn:  64726
    The number of tid with same type:  46537
    The number of unique tid:  64726

### Groupby uid
    The number of uid with same szrid:  4063
    The number of uid with diff szrid:  134
    The number of unique uid:  4197
    
    The number of uid with same churn:  4197
    The number of uid with diff churn:  0
    The number of unique uid:  4197

    The number of uid with same tid:  293
    The number of uid with diff tid:  3904
    The number of unique uid:  4197

### Groupby szrid
    The number of szrid with same uid:  4359
    The number of szrid with diff uid:  0        
    The number of unique szrid:  4359

    The number of szrid with same tid:  347
    The number of szrid with diff tid:  4012
    The number of unique szrid:  4359

### Conclusion
    uid  >  szrid  >  tid               > logs
    4197 >  4359   >  64726 w/o 'Unset' > 182000

##### Appendix

In [83]:
'''
# (Optional tid by log type 'vh')
for tid in tid_rmv_tid_Uns_lst[0:10]:
    if any(tid_rmv_tid_Uns_grouped.get_group(tid)['type']=='vh'):
        print tid_rmv_tid_Uns_grouped.get_group(tid)[tid_rmv_tid_Uns_grouped.get_group(tid)['type']=='vh']
        print '=='*40
        print ''
    #print tid_rmv_tid_Uns_grouped.get_group(tid)['type']
    #print type(tid_rmv_tid_Uns_grouped.get_group(tid))
    #print tid_rmv_tid_Uns_grouped.get_group(tid)['type']=='vh'
    #print tid_rmv_tid_Uns_grouped.get_group(tid)[tid_rmv_tid_Uns_grouped.get_group(tid)['type']=='vh']
    #print '=='*40
    #print ''
'''

"\n# (Optional tid by log type 'vh')\nfor tid in tid_rmv_tid_Uns_lst[0:10]:\n    if any(tid_rmv_tid_Uns_grouped.get_group(tid)['type']=='vh'):\n        print tid_rmv_tid_Uns_grouped.get_group(tid)[tid_rmv_tid_Uns_grouped.get_group(tid)['type']=='vh']\n        print '=='*40\n        print ''\n    #print tid_rmv_tid_Uns_grouped.get_group(tid)['type']\n    #print type(tid_rmv_tid_Uns_grouped.get_group(tid))\n    #print tid_rmv_tid_Uns_grouped.get_group(tid)['type']=='vh'\n    #print tid_rmv_tid_Uns_grouped.get_group(tid)[tid_rmv_tid_Uns_grouped.get_group(tid)['type']=='vh']\n    #print '=='*40\n    #print ''\n"

d = {'one' : [1., 2., 3., 4.],  'two' : [1., 1., 1., 1.]}

rand = np.random.RandomState(1)
df = pd.DataFrame({'A': ['foo', 'bar'] * 3,
                   'B': rand.randn(6),
                   'C': rand.randint(0, 20, 6)})
 
print(df)
 
grouped = df.groupby(['A'])

print ''
print(grouped.get_group('foo'))
 
print ''
print(dict(list(grouped))['foo'])
 
print ''
print(df.loc[grouped.groups['foo']])

##### Continuous variables

In [84]:
#dfs['drt'] = dfs['drt'].fillna(0).astype(int)

In [85]:
### set index
#dfs.set_index('ts2', inplace=True) 

In [86]:
#dfs.head()

In [87]:
#continuous_var = ['drt','tBuf','tIBuf','tLBuf','tPld','tVH']

In [88]:
'''
def get_hist_and_count(var_name):
    var_name = dfs[var_name].fillna(0).astype(int)
    print 'number of 0s: ',len(var_name[var_name==0])
    print 'number of non 0s: ',len(var_name[var_name!=0])
    var_name.hist(bins=100)
'''

"\ndef get_hist_and_count(var_name):\n    var_name = dfs[var_name].fillna(0).astype(int)\n    print 'number of 0s: ',len(var_name[var_name==0])\n    print 'number of non 0s: ',len(var_name[var_name!=0])\n    var_name.hist(bins=100)\n"

In [89]:
# duration of content in ms unit.
#get_hist_and_count('drt')

In [90]:
#get_hist_wo_NaNs('drt')

In [91]:
# Re-Buffering Time
#get_hist_and_count('tBuf')

In [92]:
# Initial Buffering Time
#get_hist_and_count('tIBuf')

In [93]:
#get_hist_wo_NaNs('tIBuf')

In [94]:
# Player Loading Time
#get_hist_and_count('tPld')

In [95]:
# Viewing Time
#get_hist_and_count('tVH')

In [96]:
# Long Buffering Time
#get_hist_and_count('tLBuf')